### PR TITLE
feat: Add secrets manager for secure credential handling

### PR DESCRIPTION
**Pull request by malhotra5**
## Summary

This PR implements a comprehensive secrets manager for the OpenHands SDK that allows secure handling of sensitive credentials in conversations and bash commands.

### Core Components
- **SecretsManager class**: Manages secrets as `dict[str, callable]` where callables retrieve secret values dynamically
- **Conversation integration**: Adds `conversation.add_secrets()` method for easy secret registration
- **Automatic detection**: Uses regex with word boundaries to find secret keys in bash commands
- **Secure injection**: Automatically injects secrets as environment variables in bash commands
- **Agent integration**: Seamlessly intercepts and modifies bash actions during execution

### Key Capabilities
- ✅ Dynamic secret retrieval through callable functions
- ✅ Automatic detection of secret keys in bash commands
- ✅ Safe environment variable injection with proper escaping
- ✅ Error handling for failed secret retrieval
- ✅ Integration with existing conversation and agent architecture
- ✅ Comprehensive unit and integration tests

## Usage Example

```python
from openhands.sdk.conversation import Conversation

# Define secret retrieval functions
def get_api_key(key: str) -> str:
    return os.getenv(key, "default-value")

def get_db_password(key: str) -> str:
    return os.getenv(key, "default-password")

# Add secrets to conversation
conversation.add_secrets({
    "API_KEY": get_api_key,
    "DB_PASSWORD": get_db_password,
})

# When agent executes bash commands containing these keys:
# Original: "curl -H 'Authorization: Bearer $API_KEY' https://api.com"
# Becomes: "export API_KEY='actual-secret-value' && curl -H 'Authorization: Bearer $API_KEY' https://api.com"
```

### Architecture
- **SecretsManager**: Core class handling secret storage and retrieval
- **ConversationState**: Extended to store secrets manager instance
- **Conversation**: Added `add_secrets()` method for user-friendly API
- **Agent**: Modified `_execute_action_events()` to intercept and modify bash commands

### Security Features
- Secrets are stored as callable functions, not raw values
- Proper shell escaping prevents injection attacks
- Graceful error handling prevents execution failures
- No secrets are logged or exposed in error messages

## Testing

- **9 SecretsManager unit tests**: Core functionality, edge cases, error handling
- **4 Conversation integration tests**: End-to-end workflow validation
- **All existing tests pass**: No breaking changes to existing functionality
- **Code quality**: Passes all pre-commit hooks (ruff, pyright, pycodestyle)

### New Files
- `openhands/sdk/conversation/secrets_manager.py` - Core secrets manager implementation
- `tests/sdk/conversation/test_secrets_manager.py` - Unit tests
- `tests/sdk/conversation/test_conversation_secrets.py` - Integration tests
- `examples/secrets_example.py` - Usage demonstration

### Modified Files
- `openhands/sdk/conversation/conversation.py` - Added `add_secrets()` method
- `openhands/sdk/conversation/state.py` - Added secrets manager storage
- `openhands/sdk/conversation/__init__.py` - Added exports
- `openhands/sdk/agent/agent.py` - Added bash command interception

## Backward Compatibility

This change is fully backward compatible:
- No existing APIs are modified
- All existing functionality remains unchanged
- New features are opt-in through the `add_secrets()` method
- No new dependencies added

## Requirements Fulfilled

1. ✅ Conversation object exposes `add_secrets` method
2. ✅ Secrets are of type `dict[str, callable]` where callable takes string and returns string
3. ✅ Looks for actions when they are about to get executed
4. ✅ If action contains secret keys, calls the callable and exports the secret into bash

This implementation provides a secure, flexible, and user-friendly way to handle sensitive credentials in OpenHands SDK conversations.

@malhotra5 can click here to [continue refining the PR](https://app.all-hands.dev/conversations/41bb840a9ab74116bdb470d0eecab7d2)

### CODE REVIEW

File: openhands/sdk/conversation/secrets_manager.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import re
+from typing import Callable, Dict
+
+from openhands.sdk.logger import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class SecretsManager:
```
Comment (xingyaoww) [inline openhands/sdk/conversation/secrets_manager.py:14]: This should be a Pydantic BaseModel from the start, so we can serialize it with the conversation

File: openhands/sdk/conversation/secrets_manager.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    """Manages secrets and injects them into bash commands when needed.
+
+    The secrets manager stores a mapping of secret keys to callable functions
+    that retrieve the actual secret values. When a bash command is about to be
+    executed, it scans the command for any secret keys and injects the corresponding
+    environment variables.
```
Comment (xingyaoww) [inline openhands/sdk/conversation/secrets_manager.py:20]: if we are not export the env var, will the agent have issue accessing these secret (eg in a python script), unless the agent explicitly export them before running the script?

If the agent is able to export them to shell, why not just export them in the first place to simplify things a bit? I think this can be just used to strip out sensitive values from the Events

Comment (malhotra5) [reply openhands/sdk/conversation/secrets_manager.py:20]: Yep exactly 

OH didnt do a good first shot at this I'll be iterating on this PR during my downtimes

Comment (malhotra5) [reply openhands/sdk/conversation/secrets_manager.py:20]: @openhands address this feedback - we should be exporting secrets and not modifying the action events themselves

don't modify or run any unit tests for now, just focus on redoing this

**Comment by malhotra5** [comment]
@openhands read the diff of this PR carefully

Then add unit tests. Before running tests make sure to remove `openhands/code` from your pythonpath

File: openhands/sdk/conversation/secrets_manager.py (hunk 1 above)
Comment (malhotra5) [reply openhands/sdk/conversation/secrets_manager.py:14]: We don't want to serialize SecretsManager because it holds callables 

It will be a private attrib instead so pydantic ignores it when serializing

File: openhands/sdk/agent/agent.py [hunk 3]
```diff
@@ -107,6 +107,31 @@ def _normalize_tools(cls, v: Any) -> dict[str, "Tool"]:
...
+                return {}
+
+        for tool in self.tools.values():
+            if (
+                tool.name == "execute_bash"
+                and hasattr(tool, "executor")
+                and tool.executor is not None
+            ):
+                # Wire the env provider for the bash executor
+                setattr(tool.executor, "env_provider", env_for_cmd)
```
Comment (xingyaoww) [inline openhands/sdk/agent/agent.py:134]: Maybe we should throw a warning when we don't have `execute_bash` Yet we have a secret manager setted?

File: openhands/sdk/conversation/state.py [hunk 4]
```diff
@@ -77,6 +83,10 @@ def assert_locked(self) -> None:
         if self._owner_tid != get_ident():
             raise RuntimeError("State not held by current thread")
 
+    def get_secrets_manager(self) -> "SecretsManager":
+        """Get the secrets manager instance."""
+        return self._secrets_manager
```
Comment (xingyaoww) [inline openhands/sdk/conversation/state.py:89]: How about 
@property 
def secret_manager

**Review by xingyaoww — APPROVED** [review]
LGTM!

I wonder if we should add a section in our system message to explain the mechanism of the secret manager: "You have access to a secret manager that will automatically export environment variables in bash when you type in relevant secrets. If you find your GITHUB_TOKEN expired, you can try to reset github remote origin by reset URL to include the latest GITHUB_TOKEN via `execute_bash` (we should show an example command here) "

**Comment by malhotra5** [comment]
> I wonder if we should add a section in our system message to explain the mechanism of the secret manager: "You have access to a secret manager that will automatically export environment variables in bash when you type in relevant secrets. If you find your GITHUB_TOKEN expired, you can try to reset github remote origin by reset URL to include the latest GITHUB_TOKEN via execute_bash (we should show an example command here) "

Yep I'll have a follow up PR which will include secrets information in the system prompt
```
 <secret name>: <secret_description>
 ```
 
 The agent doesn't need to know about how we automatically exporting envs though. Agent just knows when I use a secret it exists and does what its meant to do

**Comment by xingyaoww** [comment]
@malhotra5 I do actually think the agent should be aware of this -- I saw multiple examples of OH claimed that GitHub token expired because it hard code the previous GitHub token in the remote URL. 

Most of the time, prompting the agent to reset the URL in git remote origin with GITHUB_TOKEN fixes the issue. But without telling agent this in system prompt, we always need to manually tell agent this in follow up message
